File: frontend/app/lib/quality-gate.mjs
export const QUALITY_THRESHOLDS_V1 = Object.freeze({
  minimumLuminance: 0.15,
  minimumFaceEdgeMargin: 0.02,
  minimumConfidence: 0.5,
  minimumFrames: 3,
  minimumObservableRatio: 0.6,
  maximumGapMs: 2500,
});

export const QUALITY_MESSAGES = Object.freeze({
  no_data: "Aún no hay datos suficientes para observar esta ventana.",
  face_absent: "No se detecta un rostro en el encuadre; puedes continuar sin activar la cámara.",
  low_illumination: "La iluminación no permite obtener una señal confiable.",
  partial_face: "El rostro aparece parcialmente fuera del encuadre.",
  occluded: "La señal facial está parcialmente cubierta o incompleta.",
  low_confidence: "La señal disponible tiene confianza insuficiente.",
  insufficient_data: "Se necesitan más muestras antes de interpretar esta ventana.",
  frame_loss: "La continuidad de la señal fue insuficiente.",
  insufficient_continuity: "La ventana no tiene suficiente señal observable.",
});

export function evaluateFrame(sample, thresholds = QUALITY_THRESHOLDS_V1) {
  const features = sample?.features;
  if (!features || features.face_present === null || features.face_present === undefined) return result(false, "no_data", null);
  if (features.face_present === 0) return result(false, "face_absent", sample?.quality?.confidence ?? null);
  if (features.luminance_mean !== null && features.luminance_mean !== undefined && features.luminance_mean < thresholds.minimumLuminance) return result(false, "low_illumination", sample?.quality?.confidence ?? null);
  if (features.face_edge_margin !== null && features.face_edge_margin !== undefined && features.face_edge_margin < thresholds.minimumFaceEdgeMargin) return result(false, "partial_face", sample?.quality?.confidence ?? null);
  if (features.pose_available === 0 && features.gaze_available === 0) return result(false, "occluded", sample?.quality?.confidence ?? null);
  const confidence = sample?.quality?.confidence ?? features.detection_confidence ?? null;
  if (confidence !== null && confidence < thresholds.minimumConfidence) return result(false, "low_confidence", confidence);
  return result(true, null, confidence);
}

export function evaluateWindow(samples, thresholds = QUALITY_THRESHOLDS_V1) {
  if (!Array.isArray(samples) || samples.length === 0) return windowResult(false, "no_data", 0, 0);
  const ordered = [...samples].sort((a, b) => Date.parse(a.captured_at) - Date.parse(b.captured_at));
  if (ordered.length < thresholds.minimumFrames) return windowResult(false, "insufficient_data", 0, ordered.length);
  for (let index = 1; index < ordered.length; index += 1) {
    if (Date.parse(ordered[index].captured_at) - Date.parse(ordered[index - 1].captured_at) > thresholds.maximumGapMs) {
      return windowResult(false, "frame_loss", 0, ordered.length);
    }
  }
  const evaluated = ordered.map((sample) => evaluateFrame(sample, thresholds));
  const observable = evaluated.filter((item) => item.observable);
  const ratio = observable.length / evaluated.length;
  if (ratio < thresholds.minimumObservableRatio) {
    const reasons = evaluated.filter((item) => item.reason).map((item) => item.reason);
    const dominant = reasons.sort((a, b) => reasons.filter((item) => item === b).length - reasons.filter((item) => item === a).length)[0];
    return windowResult(false, dominant || "insufficient_continuity", ratio, evaluated.length);
  }
  const confidences = observable.map((item) => item.confidence).filter((value) => value !== null);
  return windowResult(true, null, ratio, evaluated.length, confidences.length ? confidences.reduce((a, b) => a + b, 0) / confidences.length : null);
}

function result(observable, reason, confidence) {
  return { observable, reason, confidence, message: reason ? QUALITY_MESSAGES[reason] : null };
}

function windowResult(allowInference, reason, observableRatio, sampleCount, confidence = null) {
  return { observable: allowInference, allow_inference: allowInference, allow_intervention: false, reason, observable_ratio: observableRatio, sample_count: sampleCount, confidence, message: reason ? QUALITY_MESSAGES[reason] : null };
}

File: ml/quality_gate.py
from collections import Counter


QUALITY_THRESHOLDS_V1 = {
    "minimum_luminance": 0.15,
    "minimum_face_edge_margin": 0.02,
    "minimum_confidence": 0.5,
    "minimum_frames": 3,
    "minimum_observable_ratio": 0.6,
    "maximum_gap_ms": 2500,
}


def evaluate_frame(sample, thresholds=QUALITY_THRESHOLDS_V1):
    features = sample.get("features") if isinstance(sample, dict) else None
    if not features or features.get("face_present") is None:
        return _frame(False, "no_data", None)
    confidence = (sample.get("quality") or {}).get("confidence", features.get("detection_confidence"))
    if features["face_present"] == 0:
        return _frame(False, "face_absent", confidence)
    luminance = features.get("luminance_mean")
    if luminance is not None and luminance < thresholds["minimum_luminance"]:
        return _frame(False, "low_illumination", confidence)
    margin = features.get("face_edge_margin")
    if margin is not None and margin < thresholds["minimum_face_edge_margin"]:
        return _frame(False, "partial_face", confidence)
    if features.get("pose_available") == 0 and features.get("gaze_available") == 0:
        return _frame(False, "occluded", confidence)
    if confidence is not None and confidence < thresholds["minimum_confidence"]:
        return _frame(False, "low_confidence", confidence)
    return _frame(True, None, confidence)


def evaluate_window(samples, thresholds=QUALITY_THRESHOLDS_V1):
    if not samples:
        return _window(False, "no_data", 0, 0)
    ordered = sorted(samples, key=lambda sample: sample["captured_at"])
    if len(ordered) < thresholds["minimum_frames"]:
        return _window(False, "insufficient_data", 0, len(ordered))
    times = [_milliseconds(sample["captured_at"]) for sample in ordered]
    if any(current - previous > thresholds["maximum_gap_ms"] for previous, current in zip(times, times[1:])):
        return _window(False, "frame_loss", 0, len(ordered))
    evaluated = [evaluate_frame(sample, thresholds) for sample in ordered]
    observable = [item for item in evaluated if item["observable"]]
    ratio = len(observable) / len(evaluated)
    if ratio < thresholds["minimum_observable_ratio"]:
        reasons = Counter(item["reason"] for item in evaluated if item["reason"])
        return _window(False, reasons.most_common(1)[0][0] if reasons else "insufficient_continuity", ratio, len(evaluated))
    confidence_values = [item["confidence"] for item in observable if item["confidence"] is not None]
    confidence = sum(confidence_values) / len(confidence_values) if confidence_values else None
    return _window(True, None, ratio, len(evaluated), confidence)


def _milliseconds(value):
    from datetime import datetime
    return int(datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1000)


def _frame(observable, reason, confidence):
    return {"observable": observable, "reason": reason, "confidence": confidence}


def _window(allow, reason, ratio, count, confidence=None):
    return {"observable": allow, "allow_inference": allow, "allow_intervention": False, "reason": reason, "observable_ratio": ratio, "sample_count": count, "confidence": confidence}
